GYR1-612 Hub editable 14-C updates ty2024 page 1

spompea-cfa opened the pull request:
## Link to Jira issue

- https://codeforamerica.atlassian.net/browse/GYR1-612

## Is PM acceptance required?

- Yes - don't merge until JIRA issue is accepted!

**Reminder**: merge main into this branch and get green tests before merging to main

## What was done?

- Generally speaking, updated p1 of the hub's editable 14-c screen for ty2024.
- Added new questions to Hub 14-C form p1:
  - Did you live or work in two or more states in 2024?
  - primary/spouse held/own and digital assets?
  - Spouse issued IPPIN?
  - Added IPPIN question for dependents as well
- Move a slew of questions from p3 to p1 (in keeping with the ty2024 pdf)
  - UPDATE: i'd made some changes to p3 code but rolled it back jan 16th, for better PR clarity.
- Added a grey background for partner-only questions

*Of note:* while the pdf uses the "you/spouse/you-and-spouse/no" pattern, I've generally stuck with the two separate yes/no questions instead. This is an easier approach for when we have two distinct database columns (a primary one and a spouse one).  

Update: additional changes made during acceptance testing:
- getting the refund options to match the ty2024 better --> this resulted in new columns and changes to form classes for a couple intake flow screens!
- Added "N/A" to the 'grey' dependent questions.
- some other small changes

## How to test?

> [!NOTE] 
> Out of scope for this ticket:
> - visa checkboxes
> - balance due checkboxes 

Using the ty2024 pdf as a guide, go field by field, and ensure that the Hub 14-c form, page 1, 
has all fields. If something is missing, let me know.

Raise layout/design concerns you might have; we may end up punting those to Anu for further
discussion.

## Screenshots (for visual changes)

- Before:

<img width="1310" alt="before 1 - Screenshot 2025-01-16 at 7 44 39 PM" src="https://github.com/user-attachments/assets/739fe72b-f4f2-49ac-87cf-999a7a15a61d" />
<img width="1312" alt="before 2 - Screenshot 2025-01-16 at 7 45 06 PM" src="https://github.com/user-attachments/assets/93dcf9bf-f820-4382-860b-1e2241e984bb" />


- After:

<img width="1350" alt="a Screenshot 2025-01-14 at 8 58 01 PM" src="https://github.com/user-attachments/assets/f15ec1f2-e166-4787-aae7-7885d09aa5af" />
<img width="1351" alt="b Screenshot 2025-01-14 at 8 58 11 PM" src="https://github.com/user-attachments/assets/4a9916ee-8ae4-4762-b565-b2f1d3c98948" />
<img width="1352" alt="c Screenshot 2025-01-14 at 8 58 22 PM" src="https://github.com/user-attachments/assets/b8716ff9-6bc5-409f-bc67-6d4f7231150c" />

embarnard commented on app/forms/hub/update_13614c_form_page1.rb:
```diff
@@ -91,6 +103,19 @@ def save
       modified_attributes[:dependents_attributes] = formatted_dependents_attributes
 
+      # why, just why? in an ideal world, these would never be nil to begin with.
+      modified_attributes[:primary_lived_or_worked_in_two_or_more_states] ||= 'unfilled'
```
Could you explain to me why we have to do this?

spompea-cfa replied:
excellent question. you probably saw (or will see) that i've brought this up in slack wtih you and Kevin to discuss further, but  the short answer is that the fields are not getting set to 0/unfilled properly, and instead they're heading all the way to the db as nils, resulting in null exceptions being thrown (they're non-null columns).

anyway let's discuss further in slack. a follow-up chore ticket might be necessary.

spompea-cfa replied:
cc @powersurge360

embarnard commented:
before is in flows? @spompea-cfa

embarnard commented on app/views/hub/clients/edit_13614c_form_page3.html.erb:
```diff
@@ -20,45 +20,13 @@
           <div class="grid__item width-one-whole">
             <%= f.cfa_select(:received_irs_letter, "6. #{t(".fields.q6_letter_from_irs")}", yes_no_options_for_select) %>
           </div>
-          <div class="grid__item width-one-whole">
```
the rest of the items need to be renumbered now there is no #7 
![Screenshot 2025-01-16 at 5 07 51 PM](https://github.com/user-attachments/assets/007fef9d-dd28-4150-b787-c4b308137468)

spompea-cfa replied:
yeah, a lot of questions on p3 of the ty2023 pdf moved to p1 on ty2024, so i started making removals from p3 here (made sense in the moment) but the real work for p3 is happening in a different ticket  (GYR1-614) and for simplicity i've backed out all p3 changes from this PR.

embarnard commented on app/views/hub/clients/edit_13614c_form_page3.html.erb:
```diff
@@ -20,45 +20,13 @@
-            <%= f.cfa_select(:register_to_vote, "7. #{t(".fields.q7_register_to_vote")}", yes_no_options_for_select) %>
-          </div>
           <div class="grid__item width-one-whole">
             <%= f.cfa_select(:demographic_english_conversation, "8. #{t(".fields.q8_speak_and_understand_english")}", [
```
```suggestion
            <%= f.cfa_select(:demographic_english_conversation, "7. #{t(".fields.q8_speak_and_understand_english")}", [
```

spompea-cfa replied:
backed out all p3 changes from this PR

embarnard commented on spec/controllers/hub/clients_controller_spec.rb:
```diff
@@ -1977,23 +1986,15 @@
     end
 
-    describe "#update_13614c_form_page3" do
+    xdescribe "#update_13614c_form_page3" do
```
why was this skipped?

spompea-cfa replied:
yeah this spec test throws more of this PG::NotNullViolation stuff. will address when i work on GYR1-614 (hub p3) + i created a chore ticket for tracking this specifically, GYR1-636

cc @powersurge360

spompea-cfa replied:
e.g.,
```
1) Hub::ClientsController updating 13614c #update_13614c_form_page3 with a signed in user updates the clients intake with the 13614c data, creates a system note, and regenerates the pdf when clients press Save
     Failure/Error: @client.intake.update(attributes_for(:intake))
     
     ActiveRecord::NotNullViolation:
       PG::NotNullViolation: ERROR:  null value in column "balance_pay_from_bank" of relation "intakes" violates not-null constraint
```

cc @powersurge360

embarnard commented on spec/features/hub/edit_13614c_spec.rb:
```diff
@@ -106,6 +118,16 @@
       expect(find_field("hub_update13614c_form_page1[spouse_birth_date_day]").value).to eq "20"
       expect(find_field("hub_update13614c_form_page1[spouse_birth_date_month]").value).to eq "10"
 
+      # FIXME - intake is nil here, tests won't work as-is
```
are these commented out because you're expecting them to work but they are not?

spompea-cfa replied:
fixed.

spompea-cfa commented:
@embarnard
> "before is in flows? also bump schema conflict @spompea-cfa"

haha i guess flows won't have that. not sure why i can't respond directly to your comment, but i rebased and added 'before' screenshots

spompea-cfa commented:
@embarnard please re-review when you can thx!

spompea-cfa commented:
close and reopen to fix heroku insatnce
